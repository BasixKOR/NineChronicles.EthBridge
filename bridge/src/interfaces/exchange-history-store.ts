export interface ExchangeHistory {
    network: string,
    tx_id: string,
    sender: string,
    recipient: string,
    timestamp: string,
    amount: number,
};

export interface IExchangeHistoryStore {
    put(history: ExchangeHistory): Promise<void>;

    transferredAmountInLast24Hours(network: string, sender: string): Promise<number>;
};
