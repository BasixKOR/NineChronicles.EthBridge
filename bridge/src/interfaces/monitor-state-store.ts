export interface IMonitorStateStore {
    store(network: string, blockHash: string, txId: string): Promise<void>;
    load(network: string): Promise<{ blockHash: string, txId: string }>;
}
