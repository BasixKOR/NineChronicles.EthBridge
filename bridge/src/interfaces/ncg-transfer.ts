import { TxId } from "../types/txid";

export interface INCGTransfer {
    transfer(address: string, amount: BigInt): Promise<TxId>;
}
