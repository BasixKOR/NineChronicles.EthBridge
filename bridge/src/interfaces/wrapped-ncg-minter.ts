import { TransactionReceipt } from "web3-core";

/**
 * A minter of {@link wrapped-ncg-token.ts#wNCGToken | Wrapped NCG} to Ethereum network.
 */
export interface IWrappedNCGMinter {
    /**
     * Mint {@link wrapped-ncg-token.ts#wNCGToken | Wrapped NCG} to [Ethereum network](https://ethereum.org/).
     * @param address The address to mint {@link wrapped-ncg-token.ts#wNCGToken | Wrapped NCG}.
     * @param amount The amount of {@link wrapped-ncg-token.ts#wNCGToken | Wrapped NCG} to mint.
     */
    mint(address: string, amount: number): Promise<TransactionReceipt>;
}
