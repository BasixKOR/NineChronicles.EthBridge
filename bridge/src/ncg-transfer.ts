import axios, { AxiosResponse } from "axios";
import { INCGTransfer } from "./interfaces/ncg-transfer";
import { TxId } from "./types/txid";

interface GraphQLRequestBody {
    operationName: string | null;
    query: string;
    variables: Object;
}

export class NCGTransfer implements INCGTransfer {
    private readonly _apiEndpoint: string;

    /**
     * The address of sender;
     */
    private readonly _address: string;

    constructor(apiEndpoint: string, address: string) {
        this._apiEndpoint = apiEndpoint;
        this._address = address;
    }

    async transfer(address: string, amount: BigInt): Promise<TxId> {
        const nextTxNonce = await this.getNextTxNonce(this._address);
        const query = `mutation TransferGold($recipient: Address!, $amount: String!, $txNonce: Long!)
        { transfer(recipient: $recipient, amount: $amount, txNonce: $txNonce) }`;
        const response = await this.graphqlRequest({
            operationName: "TransferGold",
            query,
            variables: {
                recipient: address,
                amount: amount.toString(),
                txNonce: nextTxNonce,
            }});

        return response.data.data.transfer;
    }

    private async getNextTxNonce(address: string): Promise<number> {
        const query = "query GetNextTxNonce($address: Address!) { nextTxNonce(address: $address) } ";
        const response = await this.graphqlRequest({
            operationName: "GetNextTxNonce",
            query,
            variables: { address, }
        });

        return response.data.data.nextTxNonce;
    }

    private async graphqlRequest(body: GraphQLRequestBody): Promise<AxiosResponse> {
        return axios.post(this._apiEndpoint, body,
        {
            headers: {
                "Content-Type": "application/json",
            },
        });
    }
}
