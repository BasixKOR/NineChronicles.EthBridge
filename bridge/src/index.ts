import Web3 from "web3";
import { init } from "@sentry/node";
import { KmsProvider } from "@planetarium/aws-kms-provider";

import { IWrappedNCGMinter } from "./interfaces/wrapped-ncg-minter";
import { INCGTransfer } from "./interfaces/ncg-transfer";
import { EthereumBurnEventMonitor } from "./monitors/ethereum-burn-event-monitor";
import { NineChroniclesTransferredEventMonitor } from "./monitors/nine-chronicles-transferred-event-monitor";
import { WrappedNCGMinter } from "./wrapped-ncg-minter";
import { wNCGTokenAbi } from "./wrapped-ncg-token";
import { HeadlessGraphQLClient } from "./headless-graphql-client";
import { ContractDescription } from "./types/contract-description";
import { IMonitorStateStore } from "./interfaces/monitor-state-store";
import { Sqlite3MonitorStateStore } from "./sqlite3-monitor-state-store";
import { WebClient } from "@slack/web-api"
import { URL } from "url";
import { Configuration } from "./configuration";
import { NCGTransferredEventObserver } from "./observers/nine-chronicles"
import { EthereumBurnEventObserver } from "./observers/burn-event-observer"
import { KMSNCGSigner } from "./kms-ncg-signer";
import { NCGKMSTransfer } from "./ncg-kms-transfer";
import Decimal from "decimal.js";

(async () => {
    const GRAPHQL_API_ENDPOINT: string = Configuration.get("GRAPHQL_API_ENDPOINT");
    const NCG_MINTER: string = Configuration.get("NCG_MINTER");
    const KMS_PROVIDER_URL: string = Configuration.get("KMS_PROVIDER_URL");
    const KMS_PROVIDER_KEY_ID: string = Configuration.get("KMS_PROVIDER_KEY_ID");
    const KMS_PROVIDER_REGION: string = Configuration.get("KMS_PROVIDER_REGION");
    const KMS_PROVIDER_AWS_ACCESSKEY: string = Configuration.get("KMS_PROVIDER_AWS_ACCESSKEY");
    const KMS_PROVIDER_AWS_SECRETKEY: string = Configuration.get("KMS_PROVIDER_AWS_SECRETKEY");
    const KMS_PROVIDER_PUBLIC_KEY: string = Configuration.get("KMS_PROVIDER_PUBLIC_KEY");
    const WNCG_CONTRACT_ADDRESS: string = Configuration.get("WNCG_CONTRACT_ADDRESS");
    const MONITOR_STATE_STORE_PATH: string = Configuration.get("MONITOR_STATE_STORE_PATH");
    const SLACK_WEB_TOKEN: string = Configuration.get("SLACK_WEB_TOKEN");
    const EXPLORER_ROOT_URL: string = Configuration.get("EXPLORER_ROOT_URL");
    const ETHERSCAN_ROOT_URL: string = Configuration.get("ETHERSCAN_ROOT_URL");
    const SENTRY_DSN: string | undefined = Configuration.get("SENTRY_DSN", false);
    if (SENTRY_DSN !== undefined) {
        init({
            dsn: SENTRY_DSN,
        });
    }

    const CONFIRMATIONS = 10;

    const monitorStateStore: IMonitorStateStore = await Sqlite3MonitorStateStore.open(MONITOR_STATE_STORE_PATH);
    const slackWebClient = new WebClient(SLACK_WEB_TOKEN);

    const headlessGraphQLCLient = new HeadlessGraphQLClient(GRAPHQL_API_ENDPOINT);
    const kmsProvider = new KmsProvider(KMS_PROVIDER_URL, {
      region: KMS_PROVIDER_REGION,
      keyIds: [KMS_PROVIDER_KEY_ID],
      credential: {
        accessKeyId: KMS_PROVIDER_AWS_ACCESSKEY,
        secretAccessKey: KMS_PROVIDER_AWS_SECRETKEY
      }
    });
    const web3 = new Web3(kmsProvider);
    const wNCGToken: ContractDescription = {
        abi: wNCGTokenAbi,
        address: WNCG_CONTRACT_ADDRESS,
    };

    if (!web3.utils.isAddress(NCG_MINTER)) {
        throw Error("NCG_MINTER variable seems invalid because it is not valid address format.");
    }

    const kmsAddresses = await kmsProvider.getAccounts();
    if(kmsAddresses.length != 1) {
      throw Error("NineChronicles.EthBridge is supported only one address.");
    }
    const kmsAddress = kmsAddresses[0];
    console.log(kmsAddress);
    const minter: IWrappedNCGMinter = new WrappedNCGMinter(web3, wNCGToken, kmsAddress);
    const signer = new KMSNCGSigner(KMS_PROVIDER_REGION, KMS_PROVIDER_KEY_ID, {
        accessKeyId: KMS_PROVIDER_AWS_ACCESSKEY,
        secretAccessKey: KMS_PROVIDER_AWS_SECRETKEY,
    });
    const derivedAddress = "0x" + web3.utils.keccak256("0x" + Buffer.from(KMS_PROVIDER_PUBLIC_KEY, "base64").toString("hex").slice(2)).slice(26);
    if (kmsAddress.toLowerCase() !== derivedAddress.toLowerCase()) {
        throw Error("KMS_PROVIDER_PUBLIC_KEY variable seems invalid because it doesn't match to address from KMS.");
    }

    const ncgKmsTransfer = new NCGKMSTransfer(
        headlessGraphQLCLient,
        kmsAddress,
        KMS_PROVIDER_PUBLIC_KEY,
        [NCG_MINTER],
        signer
    );

    const ethereumBurnEventObserver = new EthereumBurnEventObserver(ncgKmsTransfer, slackWebClient, monitorStateStore, EXPLORER_ROOT_URL, ETHERSCAN_ROOT_URL);
    const ethereumBurnEventMonitor = new EthereumBurnEventMonitor(web3, wNCGToken, await monitorStateStore.load("ethereum"), CONFIRMATIONS);
    ethereumBurnEventMonitor.attach(ethereumBurnEventObserver);

    const ncgExchangeFeeRatio = new Decimal(0.01);  // 1%
    const ncgTransferredEventObserver = new NCGTransferredEventObserver(ncgKmsTransfer, minter, slackWebClient, monitorStateStore, EXPLORER_ROOT_URL, ETHERSCAN_ROOT_URL, ncgExchangeFeeRatio);
    const nineChroniclesMonitor = new NineChroniclesTransferredEventMonitor(await monitorStateStore.load("nineChronicles"), headlessGraphQLCLient, kmsAddress);
    nineChroniclesMonitor.attach(ncgTransferredEventObserver);

    ethereumBurnEventMonitor.run();
    nineChroniclesMonitor.run();
})().catch(error => {
    console.error(error);
    process.exit(-1);
});
