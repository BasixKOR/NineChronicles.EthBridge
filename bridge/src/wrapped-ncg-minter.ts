import Web3 from "web3";
import { TransactionReceipt, TransactionConfig } from "web3-core";
import { Contract } from "web3-eth-contract";
import Decimal from "decimal.js";
import { PromiEvent } from "web3-core";

import { ContractDescription } from "./types/contract-description";
import { IWrappedNCGMinter } from "./interfaces/wrapped-ncg-minter";
import { IGasPricePolicy } from "./policies/gas-price";

export class WrappedNCGMinter implements IWrappedNCGMinter {
    private readonly _web3: Web3;
    private readonly _contractDescription: ContractDescription;
    private readonly _contract: Contract;
    private readonly _minterAddress: string;
    private readonly _gasPricePolicy: IGasPricePolicy;
    private readonly _priorityFee: Decimal;

    /**
     *
     * @param web3
     * @param contractDescription
     * @param minterAddress
     * @param gasTipRatio Percentage of gas tips to be incorporated into the block but not X %. If you want 150%, you should pass 1.5 decimal instance.
     */
    constructor(
        web3: Web3,
        contractDescription: ContractDescription,
        minterAddress: string,
        gasPricePolicy: IGasPricePolicy,
        priorityFee: Decimal
    ) {
        this._web3 = web3;
        this._contractDescription = contractDescription;
        this._contract = new this._web3.eth.Contract(
            this._contractDescription.abi,
            this._contractDescription.address
        );
        this._minterAddress = minterAddress;
        this._gasPricePolicy = gasPricePolicy;
        this._priorityFee = priorityFee;
    }

    async mint(address: string, amount: Decimal): Promise<string> {
        //NOTICE: This can be a problem if the number of digits in amount exceeds 9e+14.
        //more detail: https://mikemcl.github.io/decimal.js/#toExpPos
        Decimal.set({ toExpPos: 900000000000000 });
        console.log(
            `Minting ${amount.toString()} ${
                this._contractDescription.address
            } to ${address}`
        );
        // e.g. '103926224184', '93574861317'
        const gasPriceString = await this._web3.eth.getGasPrice();
        const gasPrice = new Decimal(gasPriceString);
        const calculatedGasPrice =
            this._gasPricePolicy.calculateGasPrice(gasPrice);
        const { transactionHash } = await this._contract.methods
            .mint(address, this._web3.utils.toBN(amount.toString()))
            .send({ from: this._minterAddress, gasPrice: calculatedGasPrice });
        return transactionHash;
    }
}
