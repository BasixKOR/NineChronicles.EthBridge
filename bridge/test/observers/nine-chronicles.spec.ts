import Decimal from "decimal.js"

import { INCGTransfer } from "../../src/interfaces/ncg-transfer";
import { IWrappedNCGMinter } from "../../src/interfaces/wrapped-ncg-minter";
import { IMonitorStateStore } from "../../src/interfaces/monitor-state-store";
import { NCGTransferredEventObserver } from "../../src/observers/nine-chronicles";
import { WebClient as SlackWebClient } from "@slack/web-api";
import { TxId } from "../../src/types/txid";
import { IExchangeHistoryStore } from "../../src/interfaces/exchange-history-store";
import { IAddressBanPolicy } from "../../src/policies/address-ban";
import { Integration } from "../../src/integrations";

jest.mock("@slack/web-api", () => {
    return {
        WebClient: jest.fn(() => {
            return {
                chat: {
                    postMessage: jest.fn()
                },
            };
        })
    }
});

describe(NCGTransferredEventObserver.name, () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    const mockNcgTransfer: jest.Mocked<INCGTransfer> = {
        transfer: jest.fn(),
    };

    const mockWrappedNcgMinter: jest.Mocked<IWrappedNCGMinter> = {
        mint: jest.fn().mockResolvedValue({
            transactionHash: "TRANSACTION-HASH"
        }),
    };

    const mockSlackWebClient = new SlackWebClient() as SlackWebClient & {
        chat: {
            postMessage: ReturnType<typeof jest.fn>
        }
    };

    const mockMonitorStateStore: jest.Mocked<IMonitorStateStore> = {
        load: jest.fn(),
        store: jest.fn(),
    };

    const exchangeFeeRatio = new Decimal(0.01);
    const mockExchangeHistoryStore: jest.Mocked<IExchangeHistoryStore> = {
        put: jest.fn(),
        transferredAmountInLast24Hours: jest.fn(),
    };

    const limitationPolicy = {
        maximum: 100000,
        minimum: 100,
    };
    const BANNED_ADDRESS = "0x47D082a115c63E7b58B1532d20E631538eaFADde";
    const addressBanPolicy: jest.Mocked<IAddressBanPolicy> = {
        isBannedAddress: jest.fn().mockImplementation(address => address === BANNED_ADDRESS),
    };

    const mockIntegration: jest.Mocked<Integration> = {
        error: jest.fn(),
    };

    const observer = new NCGTransferredEventObserver(mockNcgTransfer, mockWrappedNcgMinter, mockSlackWebClient, mockMonitorStateStore, mockExchangeHistoryStore, "https://explorer.libplanet.io/9c-internal", "https://ropsten.etherscan.io", exchangeFeeRatio, limitationPolicy, addressBanPolicy, mockIntegration);

    describe(NCGTransferredEventObserver.prototype.notify.name, () => {
        beforeEach(() => {
            jest.clearAllMocks();
        });

        it("should record the block hash even if there is no events", () => {
            observer.notify({
                blockHash: "BLOCK-HASH",
                events: [],
            });

            expect(mockMonitorStateStore.store).toHaveBeenCalledWith("nineChronicles", {
                blockHash: "BLOCK-HASH",
                txId: null,
            });
        });

        it("shouldn't do anything if the amount is zero", async () => {
            mockExchangeHistoryStore.transferredAmountInLast24Hours.mockResolvedValueOnce(0);
            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [{
                    amount: "0",
                    blockHash: "BLOCK-HASH",
                    txId: "TX-ID",
                    memo: "0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
                    recipient: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                    sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                }],
            });

            expect(mockMonitorStateStore.store).toHaveBeenCalledWith("nineChronicles", {
                blockHash: "BLOCK-HASH",
                txId: "TX-ID",
            });

            expect(mockNcgTransfer.transfer).not.toHaveBeenCalled();
            expect(mockWrappedNcgMinter.mint).not.toHaveBeenCalled();
        });

        it("shouldn skip if the sender is banned", async () => {
            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [{
                    amount: "0",
                    blockHash: "BLOCK-HASH",
                    txId: "TX-ID",
                    memo: "0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
                    recipient: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                    sender: BANNED_ADDRESS,
                }],
            });

            expect(mockMonitorStateStore.store).toHaveBeenCalledWith("nineChronicles", {
                blockHash: "BLOCK-HASH",
                txId: null,
            });

            expect(mockNcgTransfer.transfer).not.toHaveBeenCalled();
            expect(mockWrappedNcgMinter.mint).not.toHaveBeenCalled();
        });

        it("should record exchange history though mint failed", async () => {
            mockExchangeHistoryStore.transferredAmountInLast24Hours.mockResolvedValueOnce(0);
            mockWrappedNcgMinter.mint.mockRejectedValueOnce(new Error());

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [
                    {
                        amount: "100.11",
                        memo: "0xa2D738C3442609d92F1C62BDF051D0385F644b8E",
                        blockHash: "BLOCK-HASH",
                        txId: "TX-ID",
                        recipient: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                    }
                ]
            });

            expect(mockMonitorStateStore.store).toHaveBeenCalledTimes(1);
            expect(mockMonitorStateStore.store).toHaveBeenNthCalledWith(1, "nineChronicles", {
                blockHash: "BLOCK-HASH",
                txId: "TX-ID",
            });

            expect(mockExchangeHistoryStore.put).toHaveBeenCalledTimes(1);
            expect(mockExchangeHistoryStore.put).toHaveBeenNthCalledWith(1, {
                amount: 100.11,
                network: "nineChronicles",
                recipient: "0xa2D738C3442609d92F1C62BDF051D0385F644b8E",
                sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                timestamp: expect.any(String),
                tx_id: "TX-ID"
            });
        })

        it("should post slack message every events", async () => {
            const amounts = new Map<string, number>();
            mockExchangeHistoryStore.put.mockImplementation(({ sender, amount }) => {
                if (!amounts.has(sender)) {
                    amounts.set(sender, amount);
                } else {
                    console.log("mockImpl", sender, amounts.get(sender)!, amount)
                    amounts.set(sender, amounts.get(sender)! + amount);
                }

                return Promise.resolve();
            });

            mockExchangeHistoryStore.transferredAmountInLast24Hours.mockImplementation((_, sender) => {
                return Promise.resolve(amounts.get(sender) || 0);
            });

            const sender = "0x2734048eC2892d111b4fbAB224400847544FC872";
            const wrappedNcgRecipient: string = "0x4029bC50b4747A037d38CF2197bCD335e22Ca301";
            function makeEvent(wrappedNcgRecipient: string, amount: string, txId: TxId) {
                return {
                    amount: amount,
                    memo: wrappedNcgRecipient,
                    blockHash: "BLOCK-HASH",
                    txId: txId,
                    recipient: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                    sender: sender,
                };
            }

            const events = [
                makeEvent(wrappedNcgRecipient, "1", "TX-INVALID-A"),
                makeEvent(wrappedNcgRecipient, "1.2", "TX-INVALID-B"),
                makeEvent(wrappedNcgRecipient, "0.01", "TX-INVALID-C"),
                makeEvent(wrappedNcgRecipient, "3.22", "TX-INVALID-D"),
                makeEvent(wrappedNcgRecipient, "100", "TX-E"),
                makeEvent(wrappedNcgRecipient, "10000000000", "TX-SHOULD-REFUND-PART-F"),
                makeEvent(wrappedNcgRecipient, "99", "TX-SHOULD-REFUND-G"),
                makeEvent(wrappedNcgRecipient, "100.01", "TX-SHOULD-REFUND-H"),
                makeEvent(wrappedNcgRecipient, "100000", "TX-SHOULD-REFUND-I"),
                makeEvent(wrappedNcgRecipient, "99999.99", "TX-SHOULD-REFUND-J"),
            ];

            console.log("HERE");

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events,
            });


            expect(mockMonitorStateStore.store).toHaveBeenNthCalledWith(1, "nineChronicles", {
                blockHash: "BLOCK-HASH",
                txId: "TX-INVALID-A",
            });

            expect(mockMonitorStateStore.store).toHaveBeenNthCalledWith(2, "nineChronicles", {
                blockHash: "BLOCK-HASH",
                txId: "TX-INVALID-B",
            });

            expect(mockMonitorStateStore.store).toHaveBeenNthCalledWith(3, "nineChronicles", {
                blockHash: "BLOCK-HASH",
                txId: "TX-INVALID-C",
            });

            expect(mockMonitorStateStore.store).toHaveBeenNthCalledWith(4, "nineChronicles", {
                blockHash: "BLOCK-HASH",
                txId: "TX-INVALID-D",
            });

            expect(mockMonitorStateStore.store).toHaveBeenNthCalledWith(5, "nineChronicles", {
                blockHash: "BLOCK-HASH",
                txId: "TX-E",
            });

            expect(mockMonitorStateStore.store).toHaveBeenNthCalledWith(6, "nineChronicles", {
                blockHash: "BLOCK-HASH",
                txId: "TX-SHOULD-REFUND-PART-F",
            });

            expect(mockMonitorStateStore.store).toHaveBeenNthCalledWith(7, "nineChronicles", {
                blockHash: "BLOCK-HASH",
                txId: "TX-SHOULD-REFUND-G",
            });
            expect(mockMonitorStateStore.store).toHaveBeenNthCalledWith(8, "nineChronicles", {
                blockHash: "BLOCK-HASH",
                txId: "TX-SHOULD-REFUND-H",
            });
            expect(mockMonitorStateStore.store).toHaveBeenNthCalledWith(9, "nineChronicles", {
                blockHash: "BLOCK-HASH",
                txId: "TX-SHOULD-REFUND-I",
            });
            expect(mockMonitorStateStore.store).toHaveBeenNthCalledWith(10, "nineChronicles", {
                blockHash: "BLOCK-HASH",
                txId: "TX-SHOULD-REFUND-J",
            });

            expect(mockExchangeHistoryStore.put).toHaveBeenNthCalledWith(1, {
                amount: 0,
                network: "nineChronicles",
                recipient: wrappedNcgRecipient,
                sender: sender,
                timestamp: expect.any(String),
                tx_id: "TX-INVALID-A"
            });

            expect(mockExchangeHistoryStore.put).toHaveBeenNthCalledWith(2, {
                amount: 0,
                network: "nineChronicles",
                recipient: wrappedNcgRecipient,
                sender: sender,
                timestamp: expect.any(String),
                tx_id: "TX-INVALID-B"
            });

            expect(mockExchangeHistoryStore.put).toHaveBeenNthCalledWith(3, {
                amount: 0,
                network: "nineChronicles",
                recipient: wrappedNcgRecipient,
                sender: sender,
                timestamp: expect.any(String),
                tx_id: "TX-INVALID-C"
            });

            expect(mockExchangeHistoryStore.put).toHaveBeenNthCalledWith(4, {
                amount: 0,
                network: "nineChronicles",
                recipient: wrappedNcgRecipient,
                sender: sender,
                timestamp: expect.any(String),
                tx_id: "TX-INVALID-D"
            });

            expect(mockExchangeHistoryStore.put).toHaveBeenNthCalledWith(5, {
                amount: 100,
                network: "nineChronicles",
                recipient: wrappedNcgRecipient,
                sender: sender,
                timestamp: expect.any(String),
                tx_id: "TX-E"
            });

            expect(mockExchangeHistoryStore.put).toHaveBeenNthCalledWith(6, {
                amount: 99900,
                network: "nineChronicles",
                recipient: wrappedNcgRecipient,
                sender: sender,
                timestamp: expect.any(String),
                tx_id: "TX-SHOULD-REFUND-PART-F"
            });

            expect(mockExchangeHistoryStore.put).toHaveBeenNthCalledWith(7, {
                amount: 0,
                network: "nineChronicles",
                recipient: wrappedNcgRecipient,
                sender: sender,
                timestamp: expect.any(String),
                tx_id: "TX-SHOULD-REFUND-G"
            });

            expect(mockExchangeHistoryStore.put).toHaveBeenNthCalledWith(8, {
                amount: 0,
                network: "nineChronicles",
                recipient: wrappedNcgRecipient,
                sender: sender,
                timestamp: expect.any(String),
                tx_id: "TX-SHOULD-REFUND-H"
            });

            expect(mockExchangeHistoryStore.put).toHaveBeenNthCalledWith(9, {
                amount: 0,
                network: "nineChronicles",
                recipient: wrappedNcgRecipient,
                sender: sender,
                timestamp: expect.any(String),
                tx_id: "TX-SHOULD-REFUND-I"
            });

            expect(mockWrappedNcgMinter.mint.mock.calls).toEqual([
                [wrappedNcgRecipient, new Decimal(    99000000000000000000)],
                [wrappedNcgRecipient, new Decimal( 98901000000000000000000)],
            ]);
        });

        for (const invalidMemo of [
            "0x",
            "",
            "0x4029bC50b4747A037d38CF2197bCD335e22Ca301a",
            "0x0000000000000000000000000000000000000000",
            // See https://github.com/planetarium/NineChronicles.EthBridge/issues/73
            "0XC1912FEE45D61C87CC5EA59DAE31190FFFFF232D",
            "0Xc1912fee45d61c87cc5ea59dae31190fffff232d",
            "C1912FEE45D61C87CC5EA59DAE31190FFFFF232D",
            null,
        ]) {
            it(`should refund with invalid memo, ${invalidMemo}`, async () => {
                await observer.notify({
                    blockHash: "BLOCK-HASH",
                    events: [
                        {
                            amount: "100.11",
                            memo: invalidMemo,
                            blockHash: "BLOCK-HASH",
                            txId: "TX-A",
                            recipient: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                            sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                        }
                    ]
                });

                expect(mockNcgTransfer.transfer).toHaveBeenCalledWith(
                    "0x2734048eC2892d111b4fbAB224400847544FC872",
                    "100.11",
                    "I'm bridge and you should transfer with memo, valid ethereum address to receive.");

                expect(mockExchangeHistoryStore.put).toHaveBeenCalledWith({
                    network: "nineChronicles",
                    tx_id: "TX-A",
                    sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                    recipient: invalidMemo ?? "",
                    timestamp: expect.any(String),
                    amount: 0,
                });
            });
        }

        it("slack message - snapshot", async () => {
            mockExchangeHistoryStore.transferredAmountInLast24Hours.mockResolvedValue(0);

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [
                    {
                        amount: "100.23",
                        memo: "0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
                        blockHash: "BLOCK-HASH",
                        txId: "TX-ID",
                        recipient: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                    },
                ],
            });

            expect(mockSlackWebClient.chat.postMessage.mock.calls).toMatchSnapshot();
        })

        it("slack refund error message - snapshot", async () => {
            mockNcgTransfer.transfer.mockImplementationOnce((address, amount, memo) => {
                throw new Error("mockNcgTransfer.transfer error");
            });

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [
                    {
                        amount: "1.23",
                        memo: "0x0000000000000000000000000000000000000000",
                        blockHash: "BLOCK-HASH",
                        txId: "TX-ID",
                        recipient: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                    },
                ],
            });

            expect(mockSlackWebClient.chat.postMessage.mock.calls).toMatchSnapshot();
        });

        it("slack object error message - snapshot", async () => {
            mockWrappedNcgMinter.mint.mockImplementationOnce(() => {
                throw {
                    code: -32000,
                    message: 'err: max fee per gas less than block base fee: address 0x9093dd96c4bb6b44A9E0A522e2DE49641F146223, maxFeePerGas: 300000000000 baseFee: 305545815494 (supplied gas 11118348)'
                };
            });

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [
                    {
                        amount: "1.23",
                        memo: "0x0000000000000000000000000000000000000000",
                        blockHash: "BLOCK-HASH",
                        txId: "TX-ID",
                        recipient: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                    },
                ],
            });

            expect(mockSlackWebClient.chat.postMessage.mock.calls).toMatchSnapshot();
        });

        it("slack ethereum transfer error message - snapshot", async () => {
            mockWrappedNcgMinter.mint.mockImplementationOnce((address, amount) => {
                throw new Error("mockWrappedNcgMinter.mint error");
            });

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [
                    {
                        amount: "100.23",
                        memo: "0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
                        blockHash: "BLOCK-HASH",
                        txId: "TX-ID",
                        recipient: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                    },
                ],
            });

            expect(mockSlackWebClient.chat.postMessage.mock.calls).toMatchSnapshot();
        });

        it("pagerduty ethereum transfer error message - snapshot", async () => {
            mockWrappedNcgMinter.mint.mockImplementationOnce((address, amount) => {
                throw new Error("mockWrappedNcgMinter.mint error");
            });

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [
                    {
                        amount: "100.23",
                        memo: "0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
                        blockHash: "BLOCK-HASH",
                        txId: "TX-ID",
                        recipient: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                    },
                ],
            });

            expect(mockIntegration.error.mock.calls).toMatchSnapshot();
        });

        // Try to catch cases when others, not object and error, were thrown.
        it("slack string error message - snapshot", async () => {
            mockWrappedNcgMinter.mint.mockRejectedValueOnce("error message");

            await observer.notify({
                blockHash: "BLOCK-HASH",
                events: [
                    {
                        amount: "1.23",
                        memo: "0x0000000000000000000000000000000000000000",
                        blockHash: "BLOCK-HASH",
                        txId: "TX-ID",
                        recipient: "0x6d29f9923C86294363e59BAaA46FcBc37Ee5aE2e",
                        sender: "0x2734048eC2892d111b4fbAB224400847544FC872",
                    },
                ],
            });

            expect(mockSlackWebClient.chat.postMessage.mock.calls).toMatchSnapshot();
        });
    })
})
