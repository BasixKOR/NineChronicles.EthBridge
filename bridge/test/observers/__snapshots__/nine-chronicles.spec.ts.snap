// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NCGTransferredEventObserver notify slack message - snapshot 1`] = `
Array [
  Array [
    Object {
      "attachments": Array [
        Object {
          "author_name": "Bridge Event",
          "color": "#42f5aa",
          "fallback": "NCG 0x2734048eC2892d111b4fbAB224400847544FC872 → wNCG 0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
          "fields": Array [
            Object {
              "title": "9c network transaction",
              "value": "https://explorer.libplanet.io/9c-internal/transaction/?TX-ID",
            },
            Object {
              "title": "Ethereum network transaction",
              "value": "https://ropsten.etherscan.io/tx/TRANSACTION-HASH?",
            },
            Object {
              "title": "sender (NineChronicles)",
              "value": "0x2734048eC2892d111b4fbAB224400847544FC872",
            },
            Object {
              "title": "recipient (Ethereum)",
              "value": "0x4029bC50b4747A037d38CF2197bCD335e22Ca301",
            },
            Object {
              "title": "amount",
              "value": "1.23",
            },
          ],
        },
      ],
      "channel": "#nine-chronicles-bridge-bot",
      "text": "NCG → wNCG event occurred.",
    },
  ],
]
`;
